llvm/include/llvm/Support/Compiler.h

```diff
@@ -157,6 +157,25 @@
 #define LLVM_DEPRECATED(MSG, FIX) [[deprecated(MSG)]]
 #endif
 
+// clang-format off
+#if defined(__clang__) || defined(__GNUC__)
+#define LLVM_IGNORE_DEPRECATIONS_OF_DECLARATIONS_BEGIN                         \
```

Comment:
Maybe we could have a more generic macro?

```c++
#define DO_PRAGMA(x) _Pragma(#x)
#if defined(__clang__) || defined(__GNUC__)
#define LLVM_DISABLE_WARNING_BEGIN(GCC_FLAG, MSVC_NUMBER) \
  DO_PRAGMA(GCC diagnostic push)                          \
  DO_PRAGMA(GCC diagnostic ignored "-W" GCC_FLAG)
#define LLVM_DISABLE_WARNING_END                          \
  DO_PRAGMA(GCC diagnostic pop)
#elif defined(_MSC_VER)
#define LLVM_DISABLE_WARNING_BEGIN(GCC_FLAG, MSVC_NUMBER) \
  DO_PRAGMA(warning(push))                                \
  DO_PRAGMA(warning(disable: MSVC_NUMBER))
#define LLVM_DISABLE_WARNING_END                          \
  DO_PRAGMA(warning(pop))
#else
#define LLVM_DISABLE_WARNING_BEGIN(GCC_FLAG, MSVC_NUMBER)
#define LLVM_DISABLE_WARNING_END
#endif

LLVM_DISABLE_WARNING_BEGIN("deprecated-declarations", 4996)
EXPECT_EQ("dir/f2.cpp", F2->getFileEntry().getName());
LLVM_DISABLE_WARNING_END
```

Reply:
I'm not sure how often the general form of this is useful - I don't actually see many places in the code that disable diagnostics for both MSVC and clang/gcc.  So I would stick with the version that is specific to deprecations.  Maybe you should more closely match the name of `_LIBCPP_SUPPRESS_DEPRECATED_PUSH`  -- e.g. `LLVM_SUPPRESS_DEPRECATED_PUSH` and `..._POP`.

Reply:
OK, I used the suggested naming, but added `DECLARATIONS` to distinguish from semantics of the libc++ version that also ignores `-Wdeprecated`, which I don't want to (and I'm not sure there's an MSVC counterpart). I left this in `Compiler.h`, since this might be useful in other parts of LLVM and for downstream projects that have extra usages of `FileEntry::getName()`.